lib/fluent/plugin/filter.rb

```diff
@@ -28,24 +28,57 @@ class Filter < Base
       include PluginLoggerMixin
       include PluginHelper::Mixin
 
+      class DuplicatedFilterMethodError < StandardError; end
+
       helpers :event_emitter
 
       def filter(tag, time, record)
         raise NotImplementedError, "BUG: filter plugins MUST implement this method"
       end
 
+      def filter_with_time(tag, time, record)
+        raise NotImplementedError, "BUG: filter plugins MUST implement this method"
+      end
+
       def filter_stream(tag, es)
         new_es = MultiEventStream.new
         es.each do |time, record|
           begin
-            filtered_record = filter(tag, time, record)
-            new_es.add(time, filtered_record) if filtered_record
+            filtered_record = do_filter(tag, time, record)
+            new_es.add(*filtered_record) if filtered_record
           rescue => e
             router.emit_error_event(tag, time, record, e)
           end
         end
         new_es
       end
+
+      private
+
+      def do_filter(tag, time, record)
```

Comment:
This method is calling `if` every time for `each` blocks.
Of course it makes a small performance overhead (and your code creates `[time, record]` instance every time, which makes one more overhead).
Please make `if` outside of `es.each` and write different code for each methods.
```
if @without_time
  es.each do |time, record|
    begin
      new_es.add(time, filter(tag, time, record))
    rescue => e
      # ..
    end
  end
else
  # ...
end
```

This code looks not so cool, but doesn't create additional objects and doesn't call additional `if` and other methods. It's very important for loops heavily called of data processing middleware.